### PR TITLE
Initial version of Garbage Collector

### PR DESCRIPTION
**Pull request by MikhailBurdukov**

### CODE REVIEW

File: src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    LOG_DEBUG(log, "Merge snapshot with {} entries from wal.", wal_items_batch.size());
+    auto entires_to_remove = mergeWithWals(wal_items_batch, *shapshot_read_buffer, *new_shapshot_write_buffer);
+
+    SnapshotMetadata new_snaphot_meta(new_object.remote_path);
+    updateShapshotMetadata(new_snaphot_meta, snapshot_meta.znode_version);
+    /// When new snaphot commited, old entries can be removed.
+    /// TODO: Here we must remove in fault-tolerance way. So if the ch-server will be restarted at this point,
+    /// old objects should be scheduled for deletion. removal.
```
Comment (MikhailBurdukov) [inline src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp]: Please, take a look.

File: src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    LOG_DEBUG(log, "GC iteration finished");
+}
+
+String VFSGarbageCollector::getZKSnapshotPath() const
+{
+    return settings.zk_gc_path + "/garbage_collector/shapshot_meta";
+}
+
+
+SnapshotMetadata VFSGarbageCollector::getCurrentSnapshotObjectPath() const
```
Comment (aalexfvk) [inline src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp]: getSnapshotMetadata() ?

File: src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    Coordination::Stat stat;
+    Coordination::Error error;
+    String content;
+
+    zookeeper->tryGet(zk_shapshot_metadata_path, content, &stat, nullptr, &error);
+
+    if (error == Coordination::Error::ZOK)
+        return SnapshotMetadata::deserialize(content, stat.version);
+    //  if no zk node with metadata yet, then it's first run of garbage collector.
+    if (error == Coordination::Error::ZNONODE)
```
Comment (aalexfvk) [inline src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp]: Suggest throwing exception

File: src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+void VFSGarbageCollector::initGCState() const
+{
+    auto zookeeper = getZookeeper();
+    auto path = settings.zk_gc_path + "/garbage_collector";
+    if (zookeeper->tryCreate(path, "", zkutil::CreateMode::Persistent) == Coordination::Error::ZOK)
+    {
+        SnapshotMetadata empty_snaphot;
+        auto [new_shapshot_write_buffer, new_object] = getShapshotWriteBufferAndSnaphotObject(empty_snaphot);
+
+        updateShapshotMetadata({new_object.remote_path}, 0);
```
Comment (aalexfvk) [inline src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp]: This should be under zk lock

File: src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+void VFSGarbageCollector::updateSnapshot() const
+{
+    SnapshotMetadata snapshot_meta = getCurrentSnapshotObjectPath();
+
+    /// For most of object stroges (like s3 or azure) we don't need the object path, it's generated randomly.
+    /// But other ones reqiested to set it manually.
+    std::unique_ptr<ReadBuffer> shapshot_read_buffer = getShapshotReadBuffer(snapshot_meta);
+    auto [new_shapshot_write_buffer, new_object] = getShapshotWriteBufferAndSnaphotObject(snapshot_meta);
```
Comment (aalexfvk) [inline src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp:182]: nit: Looks like a separate snapshot class would be useful.
And semantically it is not clear here, that brand new snapshot is created.

File: src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+    auto [wal_items_batch, max_entry_index] = getWalItems(alog, settings.batch_size);
+
+    if (wal_items_batch.size() == 0ul)
+    {
+        LOG_DEBUG(log, "Merge snapshot exit due to empty wal.");
```
Comment (aalexfvk) [inline src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp:188]: return ?

File: src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    LOG_DEBUG(log, "Merge snapshot with {} entries from wal.", wal_items_batch.size());
+    auto entires_to_remove = mergeWithWals(wal_items_batch, *shapshot_read_buffer, *new_shapshot_write_buffer);
+
+    SnapshotMetadata new_snaphot_meta(new_object.remote_path);
+    updateShapshotMetadata(new_snaphot_meta, snapshot_meta.znode_version);
+    /// When new snaphot commited, old entries can be removed.
+    /// TODO: Here we must remove in fault-tolerance way. So if the ch-server will be restarted at this point,
+    /// old objects should be scheduled for deletion. removal.
+    {
+        removeShapshotEntires(entires_to_remove);
```
Comment (aalexfvk) [inline src/Disks/ObjectStorages/VFS/VFSGarbageCollector.cpp]: I believe this line should be done before `updateShapshotMetadata()`, to keep state in snapshot consistent with state on object storage. 
But `dropUpTo()` after, it is ok.

**Pull request merged**